pxr/usd/usd/prim.h

```diff
@@ -2379,8 +2312,138 @@ class UsdPrimSiblingIterator {
+/// pair of UsdPrimSiblingIterator s, denoting a half-open range of UsdPrim
+/// siblings.  It provides a subset of container-like API, such as begin(),
+/// end(), front(), empty(), etc.
+class UsdPrimSiblingRange {
+public:
+    /// Iterator type.
+    typedef UsdPrimSiblingIterator iterator;
+    /// Const iterator type.
+    typedef UsdPrimSiblingIterator const_iterator;
+    /// Iterator difference type.
+    typedef std::ptrdiff_t difference_type;
+    /// Iterator value_type.
+    typedef iterator::value_type value_type;
+    /// Iterator reference_type.
+    typedef iterator::reference reference;
+
+    UsdPrimSiblingRange() = default;
+    UsdPrimSiblingRange(const UsdPrimSiblingRange&) = default;
+    UsdPrimSiblingRange(UsdPrimSiblingRange&&) = default;
+    UsdPrimSiblingRange& operator=(const UsdPrimSiblingRange&) = default;
+    UsdPrimSiblingRange& operator=(UsdPrimSiblingRange&&) = default;
+
+    /// Construct with a pair of iterators.
+    UsdPrimSiblingRange(UsdPrimSiblingIterator begin,
+                        UsdPrimSiblingIterator end) : _begin(begin),
+                                                      _end(end) {}
+
+    /// Construct/convert from another compatible range type.
+    template <class ForwardRange>
+    explicit UsdPrimSiblingRange(const ForwardRange &r) :
+        UsdPrimSiblingRange(std::begin(r), std::end(r)) {}
+
+    /// Assign from another compatible range type.
+    template <class ForwardRange>
+    UsdPrimSiblingRange &operator=(const ForwardRange &r) {
+        *this = UsdPrimSiblingRange(r);
+        return *this;
+    }
+
+    /// First iterator.
+    iterator begin() const { return _begin; }
+
+    /// First iterator.
+    const_iterator cbegin() const { return _begin; }
+
+    /// Past-the-end iterator.
+    iterator end() const { return _end; }
+
+    /// Past-the-end iterator.
+    const_iterator cend() const { return _end; }
+
+    /// Return !empty().
+    explicit operator bool() const { return !empty(); }
+
+    /// Equality compare.
+    bool equal(const UsdPrimSiblingRange& other) const {
+        return _begin == other._begin && _end == other._end;
+    }
+
+    /// Return *begin().  This range must not be empty.
+    reference front() const {
+        TF_DEV_AXIOM(!empty());
+        return *begin();
+    }
+
+    /// Advance this range's begin iterator.
+    UsdPrimSiblingRange& advance_begin(difference_type n) {
+        std::advance(_begin, n);
+        return *this;
+    }
+
+    /// Advance this range's end iterator.
+    UsdPrimSiblingRange& advance_end(difference_type n) {
+        std::advance(_end, n);
+        return *this;
+    }
+
+    /// Return begin() == end().
+    bool empty() const { return begin() == end(); }
+
+private:
+    /// Equality comparison.
+    friend bool operator==(const UsdPrimSiblingRange &lhs,
+                           const UsdPrimSiblingRange &rhs) {
+        return lhs.equal(rhs);
+    }
+
+    /// Equality comparison.
```

Comment:
I'm also wondering if the template comparision operators are really needed (though they seem slightly less unneeded than the templated constructors.